docs/grammar.rst

```diff
@@ -0,0 +1,256 @@
+
+Match zero or more occurrences of e.
+
+::
+
+       rule_name: (e1 e2)*
+
+.. _e-2:
+
+``e+``
+~~~~~~
+
+Match one or more occurrences of e.
+
+::
+
+       rule_name: (e1 e2)+
+
+``s.e+``
+~~~~~~~~
+
+Match one or more occurrences of e, separated by s. The generated parse
+tree does not include the separator. This is otherwise identical to
+``(e (s e)*)``.
+
+::
+
+       rule_name: ','.e+
+
+.. _e-3:
+
+``&e``
+~~~~~~
+
+Succeed if e can be parsed, without consuming any input.
+
+.. _e-4:
+
+``!e``
+~~~~~~
+
+Fail if e can be parsed, without consuming any input.
+
+An example taken from the Python grammar specifies that a primary
+consists of an atom, which is not followed by a ``.`` or a ``(`` or a
+``[``:
+
+::
+
+       primary: atom !'.' !'(' !'['
+
+``~``
+~~~~~
+
+Commit to the current alternative, even if it fails to parse.
+
+::
+
+       rule_name: '(' ~ some_rule ')' | some_alt
+
+In this example, if a left parenthesis is parsed, then the other
+alternative won’t be considered, even if some_rule or ‘)’ fail to be
+parsed.
+
+Left recursion
+--------------
+
+PEG parsers normally do not support left recursion but Pegen implements
+a technique that allows left recursion using the memoization cache. This
+allows us to write not only simple left-recursive rules but also more
+complicated rules that involve indirect left-recursion like
+
+::
+
+     rule1: rule2 | 'a'
+     rule2: rule3 | 'b'
+     rule3: rule1 | 'c'
+
+and “hidden left-recursion” like::
+
+::
+
+     rule: 'optional'? rule '@' some_other_rule
+
+Variables in the Grammar
+------------------------
+
+A sub-expression can be named by preceding it with an identifier and an
+``=`` sign. The name can then be used in the action (see below), like
+this: ::
+
+::
+
+       rule_name[return_type]: '(' a=some_other_rule ')' { a }
+
+Grammar actions
+---------------
+
+To avoid the intermediate steps that obscure the relationship between
+the grammar and the AST generation the PEG parser allows directly
+generating AST nodes for a rule via grammar actions. Grammar actions are
+language-specific expressions that are evaluated when a grammar rule is
+successfully parsed. These expressions can be written in Python. As an
+example of a grammar with Python actions, the piece of the parser
+generator that parses grammar files is bootstrapped from a meta-grammar
+file with Python actions that generate the grammar tree as a result of
+the parsing.
+
+In the specific case of the PEG grammar for Python, having actions
+allows directly describing how the AST is composed in the grammar
+itself, making it more clear and maintainable. This AST generation
+process is supported by the use of some helper functions that factor out
+common AST object manipulations and some other required operations that
+are not directly related to the grammar.
+
+To indicate these actions each alternative can be followed by the action
+code inside curly-braces, which specifies the return value of the
+alternative
```

Comment:
One thing that confused me at first, so could use documentation here, is that the thing in braces must be a Python expression; it can't be multiple lines of Python code.  This is sort of implied by it being `return`ed, but I think it would help to be explicit. Another possible interpretation (and possibly a useful feature in the future) would be to allow multiple statements which include a `return`.

Also missing a period.

Reply:
I think this comment is the only one remaining.  Here's a possible wording:

```
To indicate these actions, each alternative can be followed by an action
inside curly-braces, which specifies a Python expression to be evaluated
and returned for the alternative:
```